integration/graphql/src/main/java/com/blazebit/persistence/integration/graphql/GraphQLEntityViewSupport.java

```diff
@@ -624,50 +602,41 @@ public void applyFetches(DataFetchingEnvironment dataFetchingEnvironment, Entity
      * @param elementRoot The element root
      */
     public void applyFetches(DataFetchingEnvironment dataFetchingEnvironment, EntityViewSetting<?, ?> setting, String elementRoot) {
-        String prefix = elementRoot == null || elementRoot.isEmpty() ? "" : elementRoot + "/";
-        StringBuilder sb = new StringBuilder();
         DataFetchingFieldSelectionSet selectionSet = dataFetchingEnvironment.getSelectionSet();
-        GraphQLFieldsContainer rootType = (GraphQLFieldsContainer) getElementType(dataFetchingEnvironment, elementRoot);
-        try {
-            OUTER:
-            for (SelectedField field : selectionSet.getFields()) {
-                String key = field.getQualifiedName();
-                if (key.length() < prefix.length()) {
+        OUTER:
+        for (SelectedField field : selectionSet.getFields()) {
+            if (!isLeaf(field.getType())) {
+                // Skip non-leaf types because these intermediate objects types would
+                // lead to fetching the whole subtree in EntityViewConfiguration#getFetches
+                continue;
+            }
+            String[] fieldParts = field.getFullyQualifiedName().split("/");
+            String[] elementRootParts = elementRoot.isEmpty() ? new String[0] : elementRoot.split("/");
+            List<String> mappedFields = new ArrayList<>();
+            for (int i = 0; i < fieldParts.length; i++) {
+                String[] fieldPartComponents = fieldParts[i].split("\\.");
```

Comment:
Did you perform any perf measurements when you originally decided on the char-by-char processing? I guess the perf penalty would be pretty negligible here. If anything deserves optimization here, we should avoid processing the same subpaths again and again.

So what if I introduced a local (request only) cache here? I think this would bring real perf benefits while not compromising the readability of the code.